daemon/info.go

```diff
@@ -208,22 +209,31 @@ func (daemon *Daemon) fillAPIInfo(v *types.Info) {
          more information: https://docs.docker.com/go/attack-surface/`
 
 	cfg := daemon.configStore
+	v.Listeners = make([]systemtypes.ListenerInfo, 0, len(cfg.Hosts))
 	for _, host := range cfg.Hosts {
 		// cnf.Hosts is normalized during startup, so should always have a scheme/proto
-		h := strings.SplitN(host, "://", 2)
-		proto := h[0]
-		addr := h[1]
-		if proto != "tcp" {
+		addr, err := url.Parse(host)
+		if err != nil {
+			v.Warnings = append(v.Warnings, fmt.Sprintf("WARNING: Unable to parse host: %s", host))
+			continue
+		}
+		if addr.Scheme != "tcp" {
+			v.Listeners = append(v.Listeners, systemtypes.ListenerInfo{Address: addr.String()})
```

Comment:
In this case, `Insecure` should be `nil` (with `omitempty`) rather than `false`?
I'd except `Insecure` to be `false` only when mTLS is fully configured.

Or maybe change the property name to `MTLS bool`?

Reply:
Yes, so I was a bit struggling with this part. Perhaps it should be changed indeed (could use input from you and others though);

- I wanted to have this information _per listener_ (as a combination of "secure" and "insecure" listeners is possible)
- Using `Insecure bool` makes it "easier" to present this on the cli side without having to implement complicated logic on the client side (we can print an `(insecure)` after the host (e.g.), and
- I picked `Insecure` to allow it to be a bit more "flexible / generic", as there may be various reasons for a listener to be "insecure" (e.g. `/var/run/docker.sock` having "world-readable" permissions), and with "adding other kind of listeners to the list" in the back of my mind.

_However_, as you pointed out perhaps it's _too_ generic? I considered a `Warning(s)` field per Listener, which could be useful to provide arbitrary "additional" information that may not make sense to include in a fully structured approach. But the `info` struct already has a global `Warnings` field, which can include such warnings, so adding _another_ `Warnings` field felt like it may be "too much" at this time (but perhaps could be considered?)

While looking at this code, I also saw that `daemon.loadListeners()` uses a slightly more complete check to determine if a listener is "secure", so there's some inconsistency. We should probably have a single approach for all of these (and if needed "go the extra mile" to make sure it's _actually_ secure (MTLS properly configured, not just "flags are set")); https://github.com/moby/moby/blob/011e1c71ff8f444a50ec0d0cefb6aa16eaabe1e2/cmd/dockerd/daemon.go#L658

Reply:
/cc @cpuguy83 @tianon @tonistiigi - could use more eyes/input on this. I think this information (Listeners) can be quite useful, but could use input on the design / implementation (see above)